### PR TITLE
fix: allow app exit when handle returns

### PR DESCRIPTION
**Pull request by haiyizxx**
## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues
- [ ] Tag type of change

### CODE REVIEW

**Review by cgorenflo — APPROVED** [review]

File: src/report.rs [hunk 1]
```diff
@@ -7,16 +7,26 @@ use error_stack::{AttachmentKind, Context, Frame, FrameKind, Report};
...
-#[error("{0}")]
-pub struct Error(String);
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("event sub failed")]
+    EventSub,
+    #[error("event processor failed")]
+    EventProcessor,
+    #[error("broadcast failed")]
+    Broadcast,
```
Comment (sammy1991106) [inline src/report.rs:17]:
```suggestion
    Broadcaster,
```
just to match the name of the actual thing

File: src/report.rs [hunk 2]
```diff
@@ -7,16 +7,26 @@ use error_stack::{AttachmentKind, Context, Frame, FrameKind, Report};
...
+    Error(String),
+}
 
 impl Error {
     pub fn new<T>(msg: T) -> Error
     where
         T: Display,
     {
-        Error(msg.to_string())
+        Error::Error(msg.to_string())
```
Comment (sammy1991106) [inline src/report.rs:29]: Is this still needed?

File: src/lib.rs [hunk 3]
```diff
@@ -165,15 +159,23 @@ impl<'a> App<'a> {
...
+        let mut res = Ok(());
+        while let Some(result) = set.join_next().await {
+            if !token.is_cancelled() {
+                token.cancel();
 
-        Ok(())
+                res = result.map_err(Error::new).into_report().and_then(|result| result);
+            }
+        }
+        res
```
Comment (sammy1991106) [inline src/lib.rs:179]:
```suggestion
        let res = match (set.join_next().await, token.is_cancelled()) {
            (Some(result), false) => {
                token.cancel();

                result.map_err(Error::new).into_report().and_then(|result| result)
            }
            (_, true) => Ok(()),
            (None, _) => panic!("all tasks exited unexpectedly"),
        };

        while (set.join_next().await).is_some() {}

        res
```

Comment (sammy1991106) [reply src/lib.rs:179]: 1. Not a fan of mutation and always avoid when possible. 2. It's clearer this way that run returns an error iff the first task exits with an error and is not cancelled.

**Review by sammy1991106 — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 3 above)
Comment (haiyizxx) [reply src/lib.rs:179]: like it

**Review by sammy1991106 — APPROVED** [review]

**Pull request merged**